src/list.cpp

```diff
@@ -65,13 +67,37 @@ void list_dir(const std::string path, std::vector<std::string> &files)
   closedir(dp);
 }
 
-void list_probes(const std::string &search)
+typedef std::tuple<std::string, std::string, std::string> uprobe_entry;
+static std::vector<uprobe_entry> uprobes;
+
+void uprobe_each(struct bcc_usdt *uprobe)
+{
+  uprobes.push_back(std::make_tuple(uprobe->provider, uprobe->name, uprobe->bin_path));
+}
+
+void list_probes(const std::string &search, int pid)
 {
   unsigned int i, j;
   std::string line, probe;
 
   // software
-  // TODO: add here
+  if (pid > 0) {
+    void *ctx = bcc_usdt_new_frompid(pid, nullptr);
+    if (ctx == nullptr) {
+      std::cerr << "failed to initialize usdt context for pid: " << pid << std::endl;
+      return;
+    }
+    bcc_usdt_foreach(ctx, uprobe_each);
+    for (auto u : uprobes) {
+      std::string provider, probe_name, bin_path;
+      std::tie(provider, probe_name, bin_path) = u;
```

Comment:
I'm not familiar with `std::tie`, but can it decompose into references rather than copying all the strings here?

Reply:
It looks like `C++17`'s auto can decompose into references, but I think the alpine builder base is on an older toolchain.

In any event, I now only `std::get` the two entries required for listing. Hopefully, that will be sufficient?